### PR TITLE
Generate shallower ASTs in pattern translation (Scala.js users, please read)

### PR DESCRIPTION
**Pull request by retronym**
## Compatibility Note

ScalaJS users please update to 1.0.0-M8+ or 0.6.28+ for compatibility with Scala 2.12.9

## Change

Given:
```
$ cat sandbox/test.scala
class Test {
  import reflect.runtime.universe._
  (null: Tree) match {
    case Literal(Constant(value)) =>
  }
}
```

Emit:

```
$ qscalac -Xprint:patmat sandbox/test.scala
[[syntax trees at end of                    patmat]] // test.scala
package <empty> {
  class Test extends scala.AnyRef {
    def <init>(): Test = {
      Test.super.<init>();
      ()
    };
    {
      case <synthetic> val x1: reflect.runtime.universe.Tree = (null: reflect.runtime.universe.Tree);
      case11(){
        if (x1.ne(null).unary_!)
          case12();
        <synthetic> val o16: Option[reflect.runtime.universe.Literal] = scala.reflect.runtime.`package`.universe.LiteralTag.unapply(x1);
        if (o16.isEmpty)
          case12();
        <synthetic> val p3: reflect.runtime.universe.Literal = o16.get;
        if (p3.ne(null).unary_!)
          case12();
        <synthetic> val o15: Option[reflect.runtime.universe.Constant] = scala.reflect.runtime.`package`.universe.Literal.unapply(p3);
        if (o15.isEmpty)
          case12();
        <synthetic> val p5: reflect.runtime.universe.Constant = o15.get;
        if (p5.ne(null).unary_!)
          case12();
        <synthetic> val o14: Option[reflect.runtime.universe.Constant] = scala.reflect.runtime.`package`.universe.ConstantTag.unapply(p5);
        if (o14.isEmpty)
          case12();
        <synthetic> val p7: reflect.runtime.universe.Constant = o14.get;
        if (p7.ne(null).unary_!)
          case12();
        <synthetic> val o13: Option[Any] = scala.reflect.runtime.`package`.universe.Constant.unapply(p7);
        if (o13.isEmpty)
          case12();
        matchEnd10(())
      };
      case12(){
        matchEnd10(throw new MatchError(x1))
      };
      matchEnd10(x: Unit){
        x
      }
    }
  }
}
```

Rather than:

```
$ scalac-ref 2.13.x -Xprint:patmat sandbox/test.scala
[[syntax trees at end of                    patmat]] // test.scala
package <empty> {
  class Test extends scala.AnyRef {
    def <init>(): Test = {
      Test.super.<init>();
      ()
    };
    {
      case <synthetic> val x1: reflect.runtime.universe.Tree = (null: reflect.runtime.universe.Tree);
      case11(){
        if (x1.ne(null))
          {
            <synthetic> val o16: Option[reflect.runtime.universe.Literal] = scala.reflect.runtime.`package`.universe.LiteralTag.unapply(x1);
            if (o16.isEmpty.unary_!)
              {
                <synthetic> val p3: reflect.runtime.universe.Literal = o16.get;
                if (p3.ne(null))
                  {
                    <synthetic> val o15: Option[reflect.runtime.universe.Constant] = scala.reflect.runtime.`package`.universe.Literal.unapply(p3);
                    if (o15.isEmpty.unary_!)
                      {
                        <synthetic> val p5: reflect.runtime.universe.Constant = o15.get;
                        if (p5.ne(null))
                          {
                            <synthetic> val o14: Option[reflect.runtime.universe.Constant] = scala.reflect.runtime.`package`.universe.ConstantTag.unapply(p5);
                            if (o14.isEmpty.unary_!)
                              {
                                <synthetic> val p7: reflect.runtime.universe.Constant = o14.get;
                                if (p7.ne(null))
                                  {
                                    <synthetic> val o13: Option[Any] = scala.reflect.runtime.`package`.universe.Constant.unapply(p7);
                                    if (o13.isEmpty.unary_!)
                                      matchEnd10(())
                                    else
                                      case12()
                                  }
                                else
                                  case12()
                              }
                            else
                              case12()
                          }
                        else
                          case12()
                      }
                    else
                      case12()
                  }
                else
                  case12()
              }
            else
              case12()
          }
        else
          case12()
      };
      case12(){
        matchEnd10(throw new MatchError(x1))
      };
      matchEnd10(x: Unit){
        x
      }
    }
  }
}
```

### CODE REVIEW

**Comment by retronym** [comment]
As discussed in #7375, in which we dealt with the symptoms of deep ASTs by avoiding use of pattern matching in some cases, without addressing the root cause.

**Comment by retronym** [comment]
```diff
$ jardiff /tmp/out-old /tmp/out-new
diff --git a/Test.class.asm b/Test.class.asm
index 385354c..ab46278 100644
--- a/Test.class.asm
+++ b/Test.class.asm
@@ -26,7 +26,9 @@
     ASTORE 3
     ALOAD 3
     INVOKEVIRTUAL scala/Option.isEmpty ()Z
-    IFNE L0
+    IFEQ L0
+    GOTO L1
+   L0
     ALOAD 3
     INVOKEVIRTUAL scala/Option.get ()Ljava/lang/Object;
     CHECKCAST scala/reflect/api/Trees$LiteralApi
@@ -40,7 +42,9 @@
     ASTORE 5
     ALOAD 5
     INVOKEVIRTUAL scala/Option.isEmpty ()Z
-    IFNE L1
+    IFEQ L2
+    GOTO L1
+   L2
     ALOAD 5
     INVOKEVIRTUAL scala/Option.get ()Ljava/lang/Object;
     CHECKCAST scala/reflect/api/Constants$ConstantApi
@@ -54,7 +58,9 @@
     ASTORE 7
     ALOAD 7
     INVOKEVIRTUAL scala/Option.isEmpty ()Z
-    IFNE L2
+    IFEQ L3
+    GOTO L1
+   L3
     ALOAD 7
     INVOKEVIRTUAL scala/Option.get ()Ljava/lang/Object;
     CHECKCAST scala/reflect/api/Constants$ConstantApi
@@ -68,25 +74,19 @@
     ASTORE 9
     ALOAD 9
     INVOKEVIRTUAL scala/Option.isEmpty ()Z
-    IFNE L3
+    IFEQ L4
+    GOTO L1
+   L4
     GETSTATIC scala/runtime/BoxedUnit.UNIT : Lscala/runtime/BoxedUnit;
     ASTORE 1
-    GOTO L4
-   L3
-    GOTO L5
-   L2
     GOTO L5
    L1
-    GOTO L5
-   L0
-    GOTO L5
-   L5
     NEW scala/MatchError
     DUP
     ALOAD 2
     INVOKESPECIAL scala/MatchError.<init> (Ljava/lang/Object;)V
     ATHROW
-   L4
+   L5
     RETURN
     MAXSTACK = 3
     MAXLOCALS = 10
diff --git a/out.log b/out.log
index f269a29..e4c6ad4 100644
--- a/out.log
+++ b/out.log
@@ -10,31 +10,21 @@
       case <synthetic> val x1: reflect.runtime.universe.Tree = (null: reflect.runtime.universe.Tree);
       case11(){
         <synthetic> val o16: Option[reflect.runtime.universe.Literal] = scala.reflect.runtime.`package`.universe.LiteralTag.unapply(x1);
-        if (o16.isEmpty.unary_!)
-          {
-            <synthetic> val p3: reflect.runtime.universe.Literal = o16.get;
-            <synthetic> val o15: Option[reflect.runtime.universe.Constant] = scala.reflect.runtime.`package`.universe.Literal.unapply(p3);
-            if (o15.isEmpty.unary_!)
-              {
-                <synthetic> val p5: reflect.runtime.universe.Constant = o15.get;
-                <synthetic> val o14: Option[reflect.runtime.universe.Constant] = scala.reflect.runtime.`package`.universe.ConstantTag.unapply(p5);
-                if (o14.isEmpty.unary_!)
-                  {
-                    <synthetic> val p7: reflect.runtime.universe.Constant = o14.get;
-                    <synthetic> val o13: Option[Any] = scala.reflect.runtime.`package`.universe.Constant.unapply(p7);
-                    if (o13.isEmpty.unary_!)
-                      matchEnd10(())
-                    else
-                      case12()
-                  }
-                else
-                  case12()
-              }
-            else
-              case12()
-          }
-        else
-          case12()
+        if (o16.isEmpty)
+          case12();
+        <synthetic> val p3: reflect.runtime.universe.Literal = o16.get;
+        <synthetic> val o15: Option[reflect.runtime.universe.Constant] = scala.reflect.runtime.`package`.universe.Literal.unapply(p3);
+        if (o15.isEmpty)
+          case12();
+        <synthetic> val p5: reflect.runtime.universe.Constant = o15.get;
+        <synthetic> val o14: Option[reflect.runtime.universe.Constant] = scala.reflect.runtime.`package`.universe.ConstantTag.unapply(p5);
+        if (o14.isEmpty)
+          case12();
+        <synthetic> val p7: reflect.runtime.universe.Constant = o14.get;
+        <synthetic> val o13: Option[Any] = scala.reflect.runtime.`package`.universe.Constant.unapply(p7);
+        if (o13.isEmpty)
+          case12();
+        matchEnd10(())
       };
       case12(){
         matchEnd10(throw new MatchError(x1))
```

**Comment by lrytz** [comment]
Very nice! Some more guarding by `YpatmatFlatAST` needed, or a test file update if that change can be done unconditionally. What would be the strategy to get this enabled by default?

**Comment by retronym** [comment]
I can't find the conversation at the moment, but when I originally floated this idea @sjrd mentioned that it would require a downstream change to ScalaJS's backend.

My idea with this patch is to give ScalaJS a compiler to work with to make that change. Maybe ScalaJS.next then could support both tree shapes, we could flip the default setting and later remove the `-Y` option.

**Comment by mkeskells** [comment]
@retronym have you considered if there is any impact on async, if async is run after patmat

It looks (from my limited knowledge) like it should make it easier for that case, but you are the expert

**Comment by retronym** [comment]
@mkeskells I ran scala-async's test suite successfully on the first commt in the PR. (which includes tests for post-patmat usage)

**Comment by sjrd** [comment]
Thanks for the opt-in flag! It will indeed require changes in the Scala.js compiler backend. I'll have a look at it sometime this week.

File: src/compiler/scala/tools/nsc/transform/patmat/MatchCodeGen.scala [hunk 1]
```diff
@@ -204,22 +204,41 @@ trait MatchCodeGen extends Interface {
...
         // returns MatchMonad[T]
         def one(res: Tree): Tree = matchEnd APPLY (res) // a jump to a case label is special-cased in typedApply
-        protected def zero: Tree = nextCase APPLY ()
+        protected final def zero: Tree = nextCase APPLY ()
+        override def ifThenElseZero(c: Tree, thenp: Tree): Tree = {
+          if (settings.YpatmatFlatAST.value) {
+            val z = zero
+            thenp match {
+              case If(c1, thenp1, elsep1) if z equalsStructure elsep1 =>
+                If(c AND c1, thenp1, elsep1) // cleaner, leaner trees
```
Comment (sjrd) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchCodeGen.scala]: Will this case still trigger under the new scheme? IIUC, then `thenp1` part will have been fully translated, including the reversal with `If(NOT(c1), zero, ...) :: thenp1`, before we match on `thenp` in an enclosing if/then/else. It seems to me that this case should instead recognize the shapes produced by the cases below. Shouldn't it?

**Comment by sjrd** [comment]
Scala.js PR at https://github.com/scala-js/scala-js/pull/3649.

File: src/compiler/scala/tools/nsc/transform/patmat/MatchCodeGen.scala (hunk 1 above)
Comment (sjrd) [reply src/compiler/scala/tools/nsc/transform/patmat/MatchCodeGen.scala]: As I mention at https://github.com/scala-js/scala-js/pull/3649#issuecomment-492628466, I am a bit more convinced that my above comment is indeed relevant.

Comment (retronym) [reply src/compiler/scala/tools/nsc/transform/patmat/MatchCodeGen.scala]: I'm pretty sure that it is dead code under the new scheme.

Here's how the motivating test case for the #4863 changes compilation over time, up to the new scheme:

https://github.com/retronym/patmat-ast/commits/master

The previous change eliminated the nested `If`s, but did so by introducing nested `Apply` in the condition: `(c1 && c2 && ... cn)`.

The new scheme does better, AST-depth-wise, with one condition in each flattened `If`.

Comment (retronym) [reply src/compiler/scala/tools/nsc/transform/patmat/MatchCodeGen.scala]: This reminds me of the recent issues we've had with string interpolation AST depth of `"a1" + "a1" + ... "a256"`. My latest proposal is to add parens to reduce depth.

We could introduce an AST node to represent reductions like these in a flat way `Reduce(op: Symbol, operations: List[Tree])`, but that would require downstream compiler phases (including in plugins) to be modified to deal with it. We could probably cheat and squeeze it into an `Apply` node as well to look like a varargs call to a synthetic method, but that wouldn't capture the by-name-ness of the operands to `&&`, which downstream phases have (hard-coded) knowledge of ([example](https://github.com/scala/scala-async/blob/2c2e205014b6e4b421e5ed124163acb79fdb8c65/src/main/scala/scala/async/internal/TransformUtils.scala#L123)).

**Comment by retronym** [comment]
Pushed test fixes and removal of `case If(c1, thenp1, elsep1) if z equalsStructure elsep1` from the new scheme.

**Review by lrytz — APPROVED** [review]

**Comment by sjrd** [comment]
https://github.com/scala-js/scala-js/pull/3649 on Scala.js' side is basically ready to merge. So unless you're going to release 2.12.9 before 2.13.0 final, it's fine to introduce the new AST structure by default, without the flag. I have tested the Scala.js PR against the first commit of this PR, where it is the default, and everything works fine.

**Comment by sjrd** [comment]
Status update: Scala.js 0.6.28 has been released, with support for the shallower ASTs generated by this PR. We'll still need a 1.x version to be released with that support.

**Comment by lrytz** [comment]
@retronym this PR seems to be ready for merging, but needs a rebase. Since Scala.js already deals with the new ASTs, should it still be opt-in?

Either way, it would be a good test to enable it by default just to see if it bootstraps and passes the tests.

**Comment by sjrd** [comment]
Oh I forgot to mention here that Scala.js 1.0.0-M8 also supports this change. So this is good on the 1.x series as well.

**Comment by retronym** [comment]
I've dropped (no more setting to control this) the last commit and rebased.

**Pull request merged**